pkg/sentry/kernel/kernel.go

```diff
@@ -143,6 +143,10 @@ type Kernel struct {
 	// to CreateProcess, and is protected by extMu.
 	globalInit *ThreadGroup
 
+	// containerInit is the thread group whose leader has ID 1 in the container's
+	// PID namespace. This is needed on restore time to recreate loader.processes
+	containerInit []*ThreadGroup
```

Comment:
`containerInit` slice is used to stitch together `Loader` state with the `Kernel` after the kernel is restored. Instead, more state could be moved under `Kernel` so it's saved with the appropiate values. You could add something like a `ContainerState` here that contains `Loader.processes` and a map from container name to index. `execID` changes from `CID string` to `containerIndex int`. Then, the loader keeps only a map between CID to index and relies on the restored map to keep the correct state.

Then the change to add `GetMetadata()` to state file, seek back to 0, and the problem with locking the state file go away.